From 9f8983cecaba13e1c01b4dcb3a756cfdc9ad4014 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 25 Jan 2021 15:30:12 +0200
Subject: [PATCH] [SW-30630] ppboot: goya2: check ack signal in reset arc sequence

Currently in reset ARC sequence, halt request bit is cleared without
checking halt ACK from ARC. As per recommendation from Si team, this
patch add change to wait for ACK from ARC for halt request and then
proceed with clearing HALT request and reset ARC.

Since udelay is used for waiting for ARC response, the initialization
of ARC is moved after udelay and pcie init.

Change-Id: I58e86e4c14f4ec52f2b825322082ca3b5871665f
---

diff --git a/gaudi/common.c b/gaudi/common.c
index 7faf702..ee0b7a6 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -248,7 +248,7 @@
 		;
 }
 
-void hl_plat_init(void) {}
+void hl_plat_early_init(void) {}
 
 
 void ppboot_log_waiting_for_pcie(void)
diff --git a/gaudi2/common.c b/gaudi2/common.c
index d40514a..2eefe7b 100644
--- a/gaudi2/common.c
+++ b/gaudi2/common.c
@@ -400,7 +400,7 @@
 
 }
 
-void hl_plat_init(void)
+void hl_plat_early_init(void)
 {
 	init_arc();
 }
diff --git a/goya/common.c b/goya/common.c
index ad62b51..08b0394 100644
--- a/goya/common.c
+++ b/goya/common.c
@@ -195,7 +195,7 @@
 
 void preboot_sync_btl(void) {}
 
-void hl_plat_init(void) {}
+void hl_plat_early_init(void) {}
 
 void image_start(u64 loadaddr)
 {
diff --git a/goya2/common.c b/goya2/common.c
index b6533ba..fb8b89a 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -345,10 +345,56 @@
 		;
 }
 
-static void init_arc(void)
+static void halt_n_reset_arc(u8 id)
 {
-	HALT_N_RESET_ARC(0);
-	HALT_N_RESET_ARC(1);
+	u32 val, halt_ack_offset, halt_req_offset;
+	u32 retry = ARC_HALT_ACK_RETRIES;
+	u32 halt_mask, halt_ack_mask;
+	u32 rst_offset;
+
+	if (id) {
+		halt_ack_offset = mmPSOC_ARC1_CFG_RUN_HALT_ACK;
+		halt_req_offset = mmPSOC_ARC1_CFG_RUN_HALT_REQ;
+		rst_offset = mmPSOC_ARC1_CFG_ARC_RST;
+		halt_mask = PSOC_ARC1_CFG_RUN_HALT_ACK_SYS_HALT_R_MASK;
+		halt_ack_mask = PSOC_ARC1_CFG_RUN_HALT_ACK_HALT_ACK_MASK;
+
+	} else {
+		halt_ack_offset = mmPSOC_ARC0_CFG_RUN_HALT_ACK;
+		halt_req_offset = mmPSOC_ARC0_CFG_RUN_HALT_REQ;
+		rst_offset = mmPSOC_ARC0_CFG_ARC_RST;
+		halt_mask = PSOC_ARC0_CFG_RUN_HALT_ACK_SYS_HALT_R_MASK;
+		halt_ack_mask = PSOC_ARC0_CFG_RUN_HALT_ACK_HALT_ACK_MASK;
+	}
+
+	val = regmap_readl(halt_ack_offset);
+	if (val & halt_mask) {
+		preboot_puts_hex("Skipping halt for ARC", id, false);
+	} else {
+		/* Halt ARC */
+		regmap_writel(BIT(1), halt_req_offset);
+		while (1) {
+			val = regmap_readl(halt_ack_offset);
+			if (val & halt_ack_mask)
+				break;
+			udelay(10);
+			retry--;
+			if (!retry) {
+				preboot_puts_hex("No ack from ARC", id, false);
+				break;
+			}
+		}
+		regmap_writel(0x0, halt_req_offset);
+	}
+
+	/* Reset ARC */
+	regmap_writel(0x1, rst_offset);
+}
+
+void hl_plat_init(void)
+{
+	halt_n_reset_arc(0);
+	halt_n_reset_arc(1);
 
 	/*
 	 * Program Reset vector address for ARC0 and ARC1 to ICCM 0x0
@@ -419,11 +465,9 @@
 	regmap_writel(val, mmPSOC_GLOBAL_CONF_AXI_DRAIN_CTRL);
 }
 
-void hl_plat_init(void)
+void hl_plat_early_init(void)
 {
 	axi_enable_drain();
-
-	init_arc();
 }
 
 void ppboot_log_waiting_for_pcie(void)
diff --git a/goya2/common.h b/goya2/common.h
index be5ae0f..25abe29 100644
--- a/goya2/common.h
+++ b/goya2/common.h
@@ -41,24 +41,13 @@
 
 #define ARC_RST_VALUE			0x110
 #define ARC_RST_VEC_ADDR		0x0
+#define ARC_HALT_ACK_RETRIES		10
 #define ARC0_MEM_SIZE			(ARC0_ICCM_SIZE + ARC0_DCCM_SIZE)
 #define ARC1_MEM_SIZE			(ARC1_ICCM_SIZE + ARC1_DCCM_SIZE)
 
 #define HL_PCIE_LINK_RETRIES		50
 #define HL_INIT_SP_SRAM_RETRIES		50
 
-/*
- * Set ARCs in halt state before reset
- * 1. Set halt bit in RUN_HALT_REQ register
- * 2. Clear halt bit in RUN_HALT_REQ register
- * 3. Set Reset bit
- */
-#define HALT_N_RESET_ARC(id)						 \
-	do {								 \
-		regmap_writel(BIT(1), mmPSOC_ARC##id##_CFG_RUN_HALT_REQ);\
-		regmap_writel(0x0, mmPSOC_ARC##id##_CFG_RUN_HALT_REQ);	 \
-		regmap_writel(0x1, mmPSOC_ARC##id##_CFG_ARC_RST);	 \
-	} while (0)
 /*
  * Set ARC in running state
  * 1. Set run bit in RUN_HALT_REQ register
diff --git a/goya2_arc/common.c b/goya2_arc/common.c
index 86f860d..646ea38 100644
--- a/goya2_arc/common.c
+++ b/goya2_arc/common.c
@@ -254,7 +254,7 @@
 	regmap_writel(ARC_RST_VALUE, mmPSOC_ARC0_CFG_ARC_RST);
 }
 
-void hl_plat_init(void)
+void hl_plat_early_init(void)
 {
 	/* Functionality used to be done in BTL */
 	init_prstn();
diff --git a/include/preboot.h b/include/preboot.h
index c2308bd..95c475b 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -91,8 +91,9 @@
 u32 hl_skip_pcie_init(void);
 void hl_uart_init(void);
 u32 hl_spi_baudr_get(void);
-void hl_plat_init(void);
+void hl_plat_early_init(void);
 void hl_ts_init(void);
+void hl_plat_init(void);
 bool hl_is_debug_mode(enum hl_debug_hw_func func);
 bool hl_is_jtag_enabled(void);
 void hl_enter_recovery(void);
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index 08ac956..3b3f471 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -222,6 +222,8 @@
 
 __weak void ppboot_pcie_drain_init(void) {}
 
+__weak void hl_plat_init(void) {}
+
 static inline void ppboot_version(void)
 {
 	preboot_puts("Ppboot " VERSION_STR "\n");
@@ -238,7 +240,7 @@
 	u64 loadaddr;
 	bool ret;
 
-	hl_plat_init();
+	hl_plat_early_init();
 	hl_init_pcie_phy_fw();
 	hl_jtag_init();
 	/*
@@ -254,6 +256,8 @@
 	ppboot_version();
 	rst_src_dump();
 
+	hl_plat_init();
+
 	preboot_delay();
 	ret = ppboot_load_preboot_img(&loadaddr);
 	if (ret)
